services/svc_googlephotos.py

```diff
@@ -139,50 +141,50 @@ def addKeywords(self, keywords):
       self.setExtras(extras)
     return result
 
-  def prepareNextItem(self, destinationFile, supportedMimeTypes, displaySize):
-    result = self.fetchImage(destinationFile, supportedMimeTypes, displaySize)
-    if result['error'] is not None:
-      # If we end up here, two things can have happened
-      # 1. All images have been shown
-      # 2. No image or data was able to download
-      # Try forgetting all data and do another run
-      self.memoryForget()
-      for file in os.listdir(self.getStoragePath()):
-        os.unlink(os.path.join(self.getStoragePath(), file))
-      result = self.fetchImage(destinationFile, supportedMimeTypes, displaySize)
-    return result
-
-  def fetchImage(self, destinationFile, supportedMimeTypes, displaySize):
-    # First, pick which keyword to use
+  def prepareNextItem(self, destinationDir, supportedMimeTypes, displaySize, randomize):
```

Comment:
It feels like this function has gotten a bit more complicated (random vs sequential plus caching). These are great ideas, but by forcing all services to implement them, the risk for bugs have gone up.

If the `base.py` implemented a method for `getNextKeyword()` which did this random vs sequental stuff, it would make it less likely that someone (ie, me 😜 ) would screw up on writing a new service or changing one that exists.

Likewise, by putting the caching logic into the diskmanager class (which essentially is the cache manager?) and simply having a `getCachedCopy()` function (which returns `None` if none exists), there is no need to call the `getImageSize` function and the flow would make a bit more sense.

Reply:
I did my best to create a general flow for selecting an album/image. 
Now if you want to add a new service that uses keywords you only need to override `getImagesFor(keyword)` and `addUrlParams`.
https://github.com/mrworf/photoframe/blob/c71f40d6a673ea9622298bd8ae632fe8a52633f6/services/base.py#L370-L392

Everything else should be handled by `prepareNextItem`-->`selectImageFromAlbum`-->`selectImage` of the baseService!

It is still possible to return service-specific error messages, for example:
https://github.com/mrworf/photoframe/blob/c71f40d6a673ea9622298bd8ae632fe8a52633f6/services/svc_googlephotos.py#L242-L250